Boost stdlib development (#83)
coq/rfcs: text/083-boost-stdlib-dev/text.md

- Title: Boost stdlib development

- Drivers: Cyril Cohen (@CohenCyril) and Pierre Roux (@proux01)

----

# Summary

This CEP is about the technical details in making the standard library
an actual library, as listed in the
[Roadmap CEP](https://github.com/coq/ceps/pull/69).

# Motivation

As part of the promotion effort around the Coq Platform, we would like to
ensure that the stdlib does not enjoy special status and that Coq can be

> used without it. We should consider stdlib components as libraries to

Reviewer: In my opinion, a better formulation would be the following:

"We would like to ensure that any contribution, any library of the community can be integrated to the standard library."

That is, the point is not to get rid of a standard library. The community does need a standard library. The point is what we put in the "standard" library. In particular, the question is that MathComp and coq-stdpp *should* join "the" standard library.

Author: But how do you do that "join" in practice? For a basic example, let's consider [ssrnat.v](https://github.com/math-comp/math-comp/blob/master/mathcomp/ssreflect/ssrnat.v):
- how do we define `<=`?
- what naming convention?
- do we need to offer a transition path to users?

Without such details, sentences like "MathComp and coq-stdpp should join the standard library" unfortunately just remain wishful thinking.

This is a CEP, it's all about technical details that can be implemented. It's good to offer alternatives, but they should be precise enough that the alternatives can be precisely compared.

Reviewer: I realize that I had skipped this question, for which I had a proposal.

> But how do you do that "join" in practice? For a basic example, let's consider [ssrnat.v](https://github.com/math-comp/math-comp/blob/master/mathcomp/ssreflect/ssrnat.v):
> 
>     * how do we define `<=`?

`Arith` has both `<=` and `<=?`. So, we already have a system of notations for talking both of a propositional `<=` and a computational `<=`, so:
- assuming that everyone agrees with using the notation `<=` for propositional and `<=?` for computational, we can easily provide in the same library theorems talking about `<=` and theorems talking about `<=?` (with names such as, respectively, `le_trans` and `leb_trans`/`leq_trans`, possibly providing several synonymous built according to various popular naming policies);
- assuming that some users may want to see only the notation `<=` for the computational `<=`, providing an option or an optionally importable module that set `<=` to be the default notation for the computational definition (as in MathComp).

>     * what naming convention?

Considering that changing habits is in general very difficult, I'm not optimistic that adopting one naming convention that suits everyone would be easy. However, providing synonymous would be easy. E.g. we could provide naming modes and every user could then activate their favorite naming convention that the nametab would take care of.

>     * do we need to offer a transition path to users?

Regarding propositional and computational `<=`, I don't see a need for a transition path since they already cohabit as of today. (But if a consensus emerges on how to write the propositional and computational "<=" and about default naming conventions for theorems about them, why not to propose them as unique "standard" conventions, rather than supporting synonymous.)

Reviewer: With Trocq, we can also reasonably consider that any lemma referring to one or the other version of `<=` could eventuallly be used in the context of the other version.

Reviewer: As for induction schemes, whatever definition is used, I believe that we should provide different schemes, and not only the one canonically deriving from the inductive definition of current `Peano.le` (even, if maybe, some work can be done to improve the current panorama, see what `Require Import Arith. Search le "_ind"` gives for instance, showing e.g. always-satisfied hypotheses `Morphisms.Proper (Morphisms.respectful eq iff) A`).

Reviewer: If the question is also about whether `<=` should be specific to `nat` (using scopes, or typing, à la Arthur, for disambiguation) or a generic notation for the projection of a general structure equipping `nat` with an order, I would say that both are possible and that both should morally work the same. I'd like to see a detailed account of the pros and cons of each approach.

Author: Preliminary note: this is a very interesting and needed dicussion but it goes far beyond the current CEP. I nevertheless answer below, despite this "off-topic" status.

> assuming that everyone agrees with using the notation <= for propositional and <=? for computational

That's already a very strong assumption. This choice of notation assumes that propositional `<=` (shorter, looking like default) is the base / default notion whereas the boolean `<=?` is a secondary / derived thing. Mathcomp's experience tend to show rather the contrary. In practice it appears that everything can be done with the boolean leq and the propositional one is seldom, if ever, used. Following this logic, `<=` should be the boolean one and it's not even clear the propositional one deserves existing.

> we could provide naming modes and every user could then activate their favorite naming convention

I'm not convinced this would be that easy:
* you need to maintain the diferent modes, which has a cost ;
* this doesn't release you from having to choose a default.

Modes can be a great solution to offer transition paths but their value as a perennial solution appears less clear to me.

> I don't see a need for a transition path

Sure, if we don't transition anything, no transition path needed ;-)

Reviewer 2: The funny thing about `Nat.le` and `Nat.leb` is that we can presumably get a consensus about the definition of `Nat.leb` (after all, there are not that many computational ways of comparing two unary integers, except for deciding which argument is the recursor). But I am not even sure the propositional definition of `Nat.le` is consensual. For example, it could just as well have been defined as
```coq
Inductive le : nat -> nat -> Prop :=
  | le_O : forall n, le O n
  | le_S : forall m n, le m n -> S m <= S n.
```
Or following Peano's definition, as
```coq
Inductive le (n : nat) : nat -> Prop := le_n : forall k, le n (k + n).
```
And I can think of a few other propositional definitions, which all have their own benefits and downsides.

Reviewer 3: I am optimistic we can converge here, as long as we are willing to separate the defining property (or multiple) and definitional computational behavior. I think what proux is getting at is that having Nat.le compute like Nat.leb would be nice, and I agree. That's not the "definition" I would want for abstract proofs, but that's fine -- we can expose an if-and-only-if lemma as the conceptual definition. IMO the same approach extends to e.g. which inductive structure to use for defining integers, and to many other cases, but not to types that are intended to be done structural recursion over while they appear in indices of other inductives.

Reviewer: > And I can think of a few other propositional definitions, which all have their own benefits and downsides.

I have even a development where we switch between 2 definitions of `le` depending on needs (one in `SProp` extended with the Yoneda trick for dependent occurrences, one inductively-designed to get a specific induction principle: `Inductive le : nat -> nat -> Type := le_refl n: n <= n | le_down {n p}: S p <= n -> p <= n`). I don't think several definitions are a problem as soon as the motivation for each of them is clear, and that they are interchangeable.

Reviewer 3: :+1: , I also consider multiple induction principle for one type to be a pretty good experience with `induction .. using ..`.

Reviewer 3: The main reason to standardize is to avoid the low-level overhead of switching between definitions to use different stdlib lemmas in a single proof (this is currently a pain with Bezout, rel_prime, gcd, and Zis_gcd).

officially recommended libraries.

In particular, we should:

- Identify consistent stdlib components that can be used independently
  from each other and that would be worth distributing as separate
  packages. Identify their maintainers and give them freedom to define
  the future of the components they maintain, in the limits set by the
  Coq Platform charter. Allow maintainers to extract stdlib components
  to maintain and evolve them outside the core Coq repository and to have
  their own release schedule and versioning scheme, in case they wish to
  do so.

- Extract the prelude + a minimum set of components that alternative
  general libraries like MathComp and coq-stdpp need as a basis.
  Make sure that this reduced core stdlib component can evolve to allow
  other libraries to use newer features of Coq (like universe polymorphism,
  SProp and primitive projections).
  Stop including any other stdlib components as part of the `coq` opam
  package and encourage maintainers of Coq packages in other distributions
  to do the same.

# Detailed design

## Mono vs Dual Repository

- In the current situation (let's call it *monorepo*), Coq and the

>   stdlib live int the same repository (i.e. https://github.com/coq/coq ).

Reviewer 4: ```suggestion
  stdlib live in the same repository (i.e. https://github.com/coq/coq ).
```

  - Coq and stdlib development teams are mostly disjoint, the two repos
    would reflect that and enable both teams to focus on their main point 
    of interest (the small intersection could of course remain active in
    both teams).
  - The current Coq repo is scary for potential new stdlib contributors:
    it contains lot of OCaml code, enforces heavy development processes
    and the few stdlib PRs that get opened anyway often don't get timely
    reviews because most Coq developers don't care about them. We thus
    likely loose most potential new contributors to the stdlib and most
    Coq users end up with the feeling that the stdlib is not something
    that can be practically improved, do their own developments on their
    side, and never contribute anything back. Having its own repo would
    make the stdlib more welcoming for newcomers and hopefully help
    get it out of this mostly-stale situation it is suffering for years.
  - Separate repos enable each team to roll out their own release schedule
    independently. Stdlib developers no longer need to wait for the next
    Coq release, neither are they forced to release every time Coq does

>     (provided the last stdlib release remains compatible with the new Coq release).

Author: No, releases contain only .v files. That's exactly the same as any other library, including all those in the platform.

  - Minor point: a separate repo wouldn't force Coq developers to compile

>     the stdlib everytime, for instance just to run the test-suite.

Author: > I think Coq developers almost always need to compile stdlib, particularly to run the test suite.

In fact only a small portion of the test-suite tests the stdlib, most of it is testing Coq itself. That point is adressed a few paragraphs below.

>  OCaml changes may break the current stdlib or other libraries--and we want to know that sooner rather than later.

As you point, nothing specific to the stdlib here. It's the same as any other lib in the CI.

>     the stdlib everytime, for instance just to run the test-suite.

Reviewer 4: Successful compilation of stdlib and other libraries is an essential test of Coq, right?

>     the stdlib everytime, for instance just to run the test-suite.

Author: Yes, that's what the CI is for, this proposal will of course not change anything with that respect.

>     the stdlib everytime, for instance just to run the test-suite.

Reviewer 4: So IMO the "minor point" isn't helpful or persuasive.

>     the stdlib everytime, for instance just to run the test-suite.

Author: It indeed seems some developers don't really care. I personally do (a bit). For instance, when I do change something in printing, I want to rerun output tests without having to recompile the whole stdlib (that doesn't test anything regarding outputs).

>     the stdlib everytime, for instance just to run the test-suite.

Reviewer 5: @proux01 we could achieve this by merging Ali's PR to port the test-suite to dune. In fact, that PR had many other benefits (and closed a lot of bugs)

This means that the scope of the standard library, adressed below, is
everything but `theories/Init` and the above files.

### Test Suite

We need to split the current `test-suite` directory of Coq between
tests that are actually testing Coq and tests that only test the
standard library, or crucially need it. A first quick look seems to
indicate that among the about 2500 tests currently in `test-suite/`,
less than 500 fall in the second category and about 100 are ssreflect
specific, while the remaining could remain in Coq without requiring
the stdlib.

### CI

Have an initial CI setup including all things using the stdlib in
current Coq CI (that is virtually the whole Coq CI currently). Coq CI
should then require that anything applying for inclusion there,
and using the stdlib, should first enter the stdlib CI, in order to
ensure that development of the standard library doesn't break Coq CI.

### Documentation

Keep the current coqdoc documentation plus the small part of the
sphinx refman. Links on Coq website will need to be updated.

### Logpath

We'll need to keep `From Coq Require ...` for a bit, at least for
backward compatibility. But it's unclear we ultimately want to keep

> `Coq` as logpath, particularly with the forthcoming renaming to Rocq.

Author: So it seems we mostly agree here.

  enable users to specifically require a part of the stdlib while making
  sure they don't accidentally use another part (and as a minor benefit,
  save a bit of installation time with source based package managers
  such as opam). A metapackage depending on all subpackages would make
  this transparent for users still wanting to use the entire library.
  Having (= version) dependencies would make this troublefree for
  developers as each subpackage will only be compatible with other
  subpackages of the exact same version (like the coq-core and coq-stdlib
  packages already are for instance).
3. Multirepo: actually move some packages to their own repo, like
  was already done with bignums.

It seems 1. is regarded as being of general interest. When 1. is done,
2. seems like not much more effort for the benefits it brings, but 3.
seems more far fetched.
Thus this CEP currently offers to do 2., then stdlib maintainers
could agree on further repo split if need be.

### Proof of concept

Here is a first proposal for packages, see this
[proof of concept](https://github.com/proux01/roots) for details and small
changes enabling these dependencies.

### Dependency graph:

> ![dependency graph](depends.png)

Reviewer 4: Somewhat tangential:  It will be easier to improve stdlib with more tooling to understand dependencies.  Dependency graphs at the package level address are only one limited aspect.  For example, you might want to know:
- what theorems, definitions, etc. does `coq-roots-relations` use from `coq-roots-classes`?
- what packages/files fiat-crypto depends on (not in stdlib, of course)?
- what items in stdlib have no dependencies in any of the packages including in CI (possible dead code)?

IMO the best way to do this is to capture and save symbol/file references during compilation and make it simple to write scripts or commands to extract the desired info.

Displaying graphs has a lot of drawbacks:
- beyond a certain size they are unreadable and hard to search
- usefully displaying the diffs between 2 graphs seems hard to impossible

Author: Indeed quite orthogonal, maybe what you are looking for is https://github.com/coq-community/coq-dpdgraph

    theories/QArith/QArith_base.v</details>
* coq-roots-field
  * The Coq standard library, `field` tactic.
  * 9 files, 3.8 kloc
  * <details><summary>contains:</summary> theories/QArith/Qfield.v,
    theories/QArith/Qround.v, theories/QArith/Qring.v,
    theories/QArith/Qpower.v, theories/QArith/Qcanon.v,
    theories/setoid_ring/Field_theory.v,
    theories/setoid_ring/Field_tac.v, theories/setoid_ring/Rings_Q.v,
    theories/setoid_ring/Field.v</details>
* coq-roots-lqa
  * The Coq standard library, `lqa` tactic.
  * 3 files, 0.4 kloc
  * <details><summary>contains:</summary>
    theories/micromega/QMicromega.v, theories/micromega/Lqa.v,
    theories/micromega/DeclConstant.v</details>
* coq-roots-qarith
  * The Coq standard library, binary rational numbers
  * Binary rational numbers made on top of the coq-roots-nzarith
    package. These enable efficient computations in arbitrary
    precision exact rational arithmetic. Namespace `Coq.QArith`. These
    are known to be difficult to use for mathematical roofs because
    there is no canonical representation (2/3 and 4/6 are not equal
    for instance). For even more efficient computation, look at the

>     coq-bignums package which uses machine integers. For mathematic

Reviewer 4: ```suggestion
    coq-bignums package which uses machine integers. For mathematical
```

Author: Thansk for the careful reading!

(either in coq-mathcomp-ssreflect or in a new coq-mathcomp-ssr
containing only that).

### Documentation

Need to document the above split in packages (could help structure the doc).

## Maintainers

We should do an open call for maintainers. Current maintainers (stdlib
team from Coq on github) will become roots maintainer, they can of
course opt out at any time.

# Drawbacks

The stdlib will no longer come bundled with Coq. It will however be
just as easy to install via the usual package managers. More notably
it will remain an integral part of the Coq Platform.

# Alternatives

Other split in multiple repositories would be possible, could still
happen in the future, maybe not needed now. The extreme opposite would
be to lean toward a huge monorepo with everything from the platform,
probably not very practical, and wouldn't bring much since the
platform already ensures uniform releases bundling everything.
To our knowledge the monorepo approach has exactly one advantage:
it is easier to backport changes from one package to an upstream package.
It also has one major drawback: it makes it harder to have the following dependencies:

> `coq-roots-A <- coq-extralib-B <- coq-roots-C`.

Reviewer 3: I think an important alternative to consider here is performing the proposed logical-level refactoring in the current repository, without the reimagining the distribution model or conceptual role of the standard library. It would still achieve the concrete benefits for users who only want to import a part of it, retain the refactoring that has already been performed for this CEP, but minimize impact on users from further refactoring between the hard-won component boundaries. If individual components do diverge or want to add external dependencies as foreseen here, they could be spun out to their own packages as needed.

I also have opinions about the motivating vision for the future of stdlib, but to keep this discussion here focused on the component structure, I intend to post a whole another CEP for that in time for the scheduled Coq-call slot.

Author: Yes, please fill the "Alternatives" section of the CEP with your proposal and what you perceive are its pros and cons.

> I also have opinions about the motivating vision for the future of stdlib, but to keep this discussion here focused on the component structure, I intend to post a whole another CEP for that in time for the scheduled Coq-call slot.

Thanks, that will definitely be the next discussion to have once the current one gets resolved, and a CEP is probably a good way to drive the discussion.

Author: And by the way, no hurry to create the next CEP before discussing the current one in Coq Call. Those are two different discussions.

Author: > How can I add something to the alternatives section?

I'd expect you, as a maintainer, to be able to simply push to the branch of the CEP. If not, please ping me so that I give you proper rights.

> If these are different discussions, would you be willing to remove the nontechnical decisions about the future of stdlib from this CEP? Otherwise the proposed discussion schedule feels a little bit backwards.

Sure, those belong to a further CEP. The goal of the current one is only to unstuck the development of the current stdlib and "Enabling the development of an evolved or different official "stdlib"". Which one of the latter options we want to choose is outside the scope of the current CEP, we should only keep all doors open.

Reviewer 3: GitHub thinks that I do not have permissions to edit coq/ceps or your branch.

  nor time to significantly contribute to the stdlib.
  
We aim at:

- Making sure that the stdlib components that people actively rely on
  are efficiently maintained by interested parties.

- Ensuring that potential new contributors to the library feel welcome.

- Enabling the development of an evolved or different official "stdlib"
  which embraces what could be agreed as current best practices for
  library development.

- Enabling the sharing of a larger basis between important general
  libraries in the ecosystem.

- Being sure that users are properly discouraged to rely on parts of the
  stdlib that are considered subpar, especially when recommended
  alternatives are available through the Coq Platform.

To this end, we would like to
ensure that the stdlib does not enjoy special status and that Coq can be
used without it. We should consider stdlib components as libraries to
advertise for their own values as part of the Coq Platform packages, but
without their historical origin, or their development and release process
being tied to Coq, (mis)leading users to consider them as the only
officially recommended libraries.

In particular, we should:

> - Identify consistent stdlib components that can be used independently
>   from each other and that would be worth distributing as separate
>   packages. Identify their maintainers and give them freedom to define
>   the future of the components they maintain, in the limits set by the
>   Coq Platform charter. Allow maintainers to extract stdlib components
>   to maintain and evolve them outside the core Coq repository and to have
>   their own release schedule and versioning scheme, in case they wish to
>   do so.

Reviewer 2: Since it has been clarified that those are not the maintainers of each components that set their own release schedule, but the maintainers of the standard library as a whole, I suggest the following modification to make things less confusing:

```suggestion
- Allow maintainers to evolve the standard library outside the core Coq repository and
  to have their own release schedule and versioning scheme, in case they wish to
  do so. Identify consistent stdlib components that can be used independently
  from each other and that would be worth distributing as separate
  packages. Identify their maintainers and give them freedom to define
  the future of the components they maintain, in the limits set by the
  Coq Platform charter. 
```

- Extract the prelude + a minimum set of components that alternative
  general libraries like MathComp and coq-stdpp need as a basis.
  Make sure that this reduced core stdlib component can evolve to allow
  other libraries to use newer features of Coq (like universe polymorphism,
  SProp and primitive projections).
  Stop including any other stdlib components as part of the `coq` opam
  package and encourage maintainers of Coq packages in other distributions

>   to do the same.

Reviewer 2: I think that the packaging story needs to be clarified. DEB-based distributions have a `libcoq-stdlib` package, while RPM-based distributions have a `coq` package, which contains the standard library (the actual binaries are in `coq-core`, which is a dependency). They also have packages that depend on those, e.g., for Flocq, `libcoq-flocq` and `flocq`, respectively. This CEP suggests to tell the maintainers of these distributions that, instead of maintaining a single package, they now have to maintain 50 different packages.

We should make sure that this extra maintenance work will not have an adverse effect on the general availability of Coq, in particular the disappearance of existing Coq packages from distributions. (This is currently one of the strong points of Coq over Lean.) For example, I know of a few universities (starting with mine), where software for teaching purpose needs only a few days to be installed if it is just one `apt` or `dnf` command away. But if there is no official package, then you better fill a formal request one month in advance to hopefully get software to be compiled and installed in time on all the computers used by students.

By the way, to a lesser extent, the packaging concern also applies to Opam, as the official repository is currently collapsing under its own weight. A working group has been put in motion to try to devise solutions to this issue, but most of the proposed solutions seem to revolve around the idea of removing a bunch of packages from the repository (or at least archiving them away from the repository). Needless to say, the sudden appearance of 50 new Opam packages is presumably not what they had in mind.

Even regarding Coq's Platform, some clarifications are needed. It currently proposes three levels of installation, "base", "full", and "extended", as well as some manual selection of packages, and Coq's standard library is part of all of these levels. This CEP begs a few questions:
- Would the standard library be removed from "base"? Or would it stay there, perhaps only partially?
- What about "full"? Would it contain the equivalent of the current standard library or only the parts that are needed by the packages installed by "full"?
- What about the manual selection of packages? Would the user have to select among the 50 packages of the current standard library or would some larger categories be defined to hide this complexity?

Author: Note that the packages are not really "different" packages, since they all depend on other packages of the exact same version, they can almost be trated as a single package. It's only a way to offer users to be explicit about the fact that they use only part of the global package. It's opt-in for users, who can keep using the metapackage with everything.
You can see @SnarkBoojum answer above about Debian

> Would the standard library be removed from "base"? Or would it stay there, perhaps only partially?

The library is currently part of "base", the current CEP doesn't offer to change that.

> What about "full"? Would it contain the equivalent of the current standard library or only the parts that are needed by the packages installed by "full"?

I guess, as the name implies "full" should include "base", again this CEP doesn't offer any change here.

> What about the manual selection of packages? Would the user have to select among the 50 packages of the current standard library or would some larger categories be defined to hide this complexity?

There is a metapackage. Not sure if something more fine grained would be needed. Ccing @msoegtrop and @rtetley in case they have a more precise opinion.

Reviewer 6: I'm the maintainer for the Coq ecosystem in Debian and the main problems I encounter are:
- upstreams insisting they won't release a new version before a privileged distribution asks them for one (Coq Platform, generally);
- upstreams not being quite live: not officially abandoned projects, but the maintainer have so much on their plates they don't really have the time to keep them current.

Because even if the packages come from different projects, they are still globally tied: a new package for Coq breaking some other seven steps away means the Coq package is considered unsuitable. Case in point, I still couldn't upload Coq 8.19.1 to Debian, because the current status of the whole is:
```
* Update Coq [4/7]
 - [X] step 1 [3/3]
   - [X] elpi 1.18.2-1
   - [X] coq 8.19.1+dfsg-1
   - [X] coq-doc 8.19.1-1
 - [-] step 2 [15/17]
   - [X] aac-tactics 8.19.0-1
   - [X] coq-bignums 9.0.0+coq8.19-1
   - [X] coq-dpdgraph 1.0+8.19-1 ?
   - [X] coq-elpi 2.1.0-1
   - [X] coq-ext-lib 0.12.1-1
   - [X] coq-hammer 1.3.2+8.19-1
   - [X] coq-hott 8.19-1
   - [X] coq-libhyps 2.0.8-2
   - [X] coq-menhirlib 20231231+ds-2
   - [X] coq-record-update 0.3.4-1
   - [X] coq-reduction-effects 0.1.5-2
   - [X] coq-stdpp 1.10.0-1
   - [ ] coq-unicoq FAILURE [[https://github.com/unicoq/unicoq/issues/90][upstream report]]
   - [X] coq-unimath 20240331-1
   - [X] flocq 4.1.4-1
   - [ ] ott FAILURE [[https://github.com/ott-lang/ott/issues/105][upstream report]]
   - [X] paramcoq 1.1.3+coq8.19-1
 - [-] step 3 [6/8]
   - [X] coq-equations 1.3-8.19-1
   - [X] coq-gappa 1.5.5-1
   - [X] coq-hierarchy-builder 1.7.0-1
   - [X] coq-iris 4.2.0-1
   - [ ] coq-math-classes FAILURE [[https://github.com/coq-community/math-classes/issues/124][upstream report]]
   - [ ] coq-mtac2 WAIT coq-unicoq
   - [X] coq-simple-io 1.9.0-1
   - [X] coqprime 8.19-1
 - [-] step 4 [1/2]
   - [X] ssreflect 2.2.0-1
   - [ ] coq-corn WAIT coq-math-classes
 - [X] step 5 [8/8]
   - [X] coq-deriving 0.2.0-2
   - [X] coq-reglang 1.2.1-2
   - [X] coq-relation-algebra 1.7.10-2
   - [X] coquelicot 3.4.1-2
   - [X] mathcomp-bigenough 1.0.1-13
   - [X] mathcomp-finmap 2.1.0-1
   - [X] mathcomp-zify 1.5.0+2.0+8.16-2
   - [X] coq-quickchick 2.0.3-1
 - [X] step 6 [6/6]
   - [X] coq-extructures 0.4.0-2
   - [X] coq-interval 4.10.0-1
   - [X] mathcomp-algebra-tactics 1.2.3-2
   - [X] mathcomp-analysis 1.0.0-2
   - [X] mathcomp-multinomials 2.2.0-2
   - [X] mathcomp-real-closed 2.0.2-2
 - [X] step 7 [1/1]
   - [X] coqeal 2.0.2-1
```

Reviewer 2: > Note that the packages are not really "different" packages, since they all depend on other packages of the exact same version, they can almost be treated as a single package.

This seems to contradict the previous goal of CEP, which explicitly says "allow maintainers to have their own release schedule and versioning scheme". As soon as various components of the standard library have differing release schedules, it becomes very difficult to treat it as a single package.

* coq-roots-classes
  * The Coq standard library, classes in addition to the ones from stdlib.
  * Most of namespace `Coq.Classes`
  * 5 files, 0.6 kloc
  * <details><summary>contains:</summary>
    Classes/Morphisms_Relations.v, Classes/RelationPairs.v,
    Classes/SetoidClass.v,
    Classes/CEquivalence.v,
    Classes/DecidableClass.v</details>
* coq-roots-bool
  * The Coq standard library, boolean arithmetic
  * Booleans (basic functions and results). Most of namespace `Coq.Bool`.
  * 4 files, 1.1 kloc
  * <details><summary>contains:</summary> Bool/IfProp.v, Bool/Bool.v,
    Bool/DecBool.v, Bool/BoolEq.v</details>
* coq-roots-structures
  * The Coq standard library, structures
  * Basic "algebraic" structures: types with decidable equality and with order.
    Most of namespace `Coq;Structures`.
    Instances can be found in the coq-roots-orders-ex package.
    More developped algebra can be found in the mathematical components library: https://github.com/math-comp/math-comp
  * 5 files, 2.0 kloc
  * <details><summary>contains:</summary> Structures/GenericMinMax.v,
    Structures/Equalities.v, Structures/Orders.v,

>     Structures/OrdersFacts.v, Structures/OrdersTac.v</details>

Reviewer 7: I've personally used them mostly around MSets and FMap

    theories/Numbers/Integer/Abstract/ZDivTrunc.v,
    theories/Numbers/Integer/Abstract/ZParity.v,
    theories/Numbers/Integer/Abstract/ZMul.v,
    theories/Numbers/Integer/Abstract/ZPow.v,
    theories/Numbers/Integer/Abstract/ZAdd.v,
    theories/Numbers/Integer/Abstract/ZDivFloor.v,
    theories/Numbers/Integer/Abstract/ZMulOrder.v,
    theories/Numbers/Integer/Abstract/ZDivEucl.v,
    theories/Numbers/Integer/Abstract/ZMaxMin.v,
    theories/Numbers/Integer/Abstract/ZGcd.v,
    theories/Numbers/Integer/Abstract/ZLt.v,
    theories/Numbers/Integer/Abstract/ZLcm.v,
    theories/Numbers/Integer/Abstract/ZBase.v,
    theories/Numbers/Integer/Abstract/ZBits.v,
    theories/ZArith/BinIntDef.v, theories/ZArith/BinInt.v,
    theories/ZArith/Zcompare.v, theories/ZArith/Zorder.v,
    theories/ZArith/Zminmax.v, theories/ZArith/Zmin.v,
    theories/ZArith/Zmax.v, theories/ZArith/Znat.v,
    theories/ZArith/ZArith_dec.v, theories/ZArith/Zabs.v,
    theories/ZArith/auxiliary.v, theories/ZArith/Zbool.v,
    theories/ZArith/Zmisc.v, theories/ZArith/Wf_Z.v,
    theories/ZArith/Zhints.v, theories/ZArith/ZArith_base.v,
    theories/ZArith/Zeven.v, theories/ZArith/Zpow_alt.v,

>     theories/ZArith/Zeuclid.v, theories/ZArith/Int.v</details>

Author: No strong opinion, but it may be good to enforce some dependencies, like for instance positive not dependeing on zarith-base.

    theories/omega/OmegaLemmas.v, theories/omega/PreOmega.v,
    theories/micromega/ZifyN.v, theories/micromega/ZifyComparison.v,
    theories/micromega/ZifyClasses.v, theories/micromega/ZifyNat.v,
    theories/micromega/ZifyPow.v, theories/micromega/ZifyBool.v,
    theories/micromega/Zify.v, theories/micromega/ZifyInst.v,
    theories/micromega/DeclConstantZ.v,
    theories/micromega/OrderedRing.v, theories/micromega/Tauto.v,
    theories/micromega/Env.v, theories/micromega/Refl.v,
    theories/micromega/ZArith_hints.v,
    theories/micromega/ZMicromega.v, theories/micromega/EnvRing.v,
    theories/micromega/Lia.v, theories/micromega/VarMap.v,
    theories/micromega/Ztac.v, theories/micromega/ZCoeff.v,
    theories/micromega/RingMicromega.v</detail>
* coq-roots-zarith
  * The Coq standard library, binary integers
  * Binary encoding of integers. This binary encoding was initially
    developped to enable efficient computations, compared to the unary
    encoding of `nat`. Proofs were then added making the types usable
    for mathematical proofs, although this was not the initial
    intent. If even-more efficient computations are needed, look at

>     the coq-roots-primitive-int package for 63 bits machine arithmetic or the

Author: Please suggest phrasing improvements if you think it's bad. The goal is to convey the idea that zarith is good for effective computations but if users need more efficient computations, there are alternatives (with the drawback of increasing the trusted code base).

    coq-bignums package for arbitrary precision machine
    arithmetic. Everything can be imported with `From Coq Require
    Import NArith` or `From Coq Require Import ZArith`. Also contains
    the migromega tactic that can be loaded with `From Coq Require
    Import Lia.`. Most of namespaces `Coq.NArith` and `Coq.ZArith`.
  * 23 files, 6.5 kloc
  * <details><summary>contains:</summary>
    theories/Numbers/Natural/Binary/NBinary.v,
    theories/ZArith/Zpower.v, theories/ZArith/Zquot.v,
    theories/ZArith/Zpow_facts.v, theories/ZArith/Zgcd_alt.v,
    theories/ZArith/Zwf.v, theories/ZArith/ZArith.v,
    theories/ZArith/Zbitwise.v,
    theories/Numbers/Integer/Binary/ZBinary.v,
    theories/Numbers/Integer/NatPairs/ZNatPairs.v,
    theories/Numbers/DecimalN.v, theories/Numbers/DecimalNat.v,
    theories/Numbers/DecimalZ.v, theories/Numbers/DecimalFacts.v,
    theories/Numbers/DecimalPos.v, theories/Numbers/HexadecimalN.v,
    theories/Numbers/HexadecimalNat.v,
    theories/Numbers/HexadecimalZ.v,
    theories/Numbers/HexadecimalFacts.v,
    theories/Numbers/HexadecimalPos.v, theories/btauto/Algebra.v,

>     theories/btauto/Reflect.v, theories/btauto/Btauto.v</details>

Author: Btauto is using ring and lia, so they do depend on zarith, at least indirectly.

* coq-roots-qarith-base
  * The Coq standard library, binary rational numbers base package.
    Use coq-roots-qarith for more functionalities with the `field`
    and `Lqa` tactics. Namespace `Coq.QArith`.
  * 4 files, 1.8 kloc
  * <details><summary>contains:</summary>theories/QArith/Qreduction.v,
    theories/QArith/Qminmax.v, theories/QArith/QOrderedType.v,
    theories/QArith/QArith_base.v</details>
* coq-roots-field
  * The Coq standard library, `field` tactic.
  * 9 files, 3.8 kloc
  * <details><summary>contains:</summary> theories/QArith/Qfield.v,
    theories/QArith/Qround.v, theories/QArith/Qring.v,
    theories/QArith/Qpower.v, theories/QArith/Qcanon.v,
    theories/setoid_ring/Field_theory.v,
    theories/setoid_ring/Field_tac.v, theories/setoid_ring/Rings_Q.v,

>     theories/setoid_ring/Field.v</details>

Reviewer 3: It is unfortunate that field comes bundled with Q. Needn't do anything about it now, just another case where package the package boundaries are not really good for external consumption even though they would be immediately useful internally.

* coq-primitive-array
  * The Coq standard library, primitive persistent arrays
  * Imperative arrays with a functionnal interface. Enables efficient
    computations with arrays. Namespace `Coq.Array`.
  * 2 files, 0.0 kloc
  * <details><summary>contains:</summary> Array/PArray.v</details>
* coq-primitive-floats
  * The Coq standard library, primitive floats
  * Interface to hardware floats for efficient computations. This
    offers a basic model of floating-point arithmetic but is not very
    usable alone. Look at the coq-flocq package for an actual model of
    floating-point arithmetic, including links to coq-roots-reals and
    the current coq-primitive-floats. Namespace `Coq.Floats`
  * 2 files, 0.1 kloc
  * <details><summary>contains:</summary> Floats/Floats.v, 
    Floats/FloatLemmas.v</details>
* coq-roots-string
  * The Coq standard library, strings
  * Implementation of string as list of ASCII characters. Namespace
    `Coq.Strings`.
  * 9 files, 3.2 kloc
  * <details><summary>contains:</summary> Strings/ByteVector.v,
    Strings/HexString.v, Strings/Ascii.v, Strings/String.v,
    Strings/OctalString.v, Strings/Byte.v, Strings/BinaryString.v,

>     Numbers/DecimalString.v, Numbers/HexadecimalString.v</details>

Author: No, you could probably try to dive into the history and ask the author of that file why it got put into the `String` directory in the first place.

>     Numbers/DecimalString.v, Numbers/HexadecimalString.v</details>

Reviewer 7: Same reason that ASCII is under strings, byte was originally added just to be efficient strings

>     Numbers/DecimalString.v, Numbers/HexadecimalString.v</details>

Reviewer 3: Hmm, maybe they do belong together in this case. Is there also a vision for Notations to use Gallina code for parsing lists of bytes, or is that intentionally spearate with decimal types and so on?

>     Numbers/DecimalString.v, Numbers/HexadecimalString.v</details>

Reviewer 7: I think right now only string notations parse lists of bytes.  I think there are some vague plans to allow custom entries to specify that strings should be delimited by something other than `"`, which would allow for arbitrary parsing from lists of bytes in custom entries, but I don't think there's currently any plan to move general notation parsing code to be based on lists of bytes / to live in Gallina.

* coq-roots-extraction
  * The Coq standard library, extraction
  * Interface to the extraction module, load with
    `From Coq Require Extraction.`. Namespace `Coq.extraction`.
  * 21 files, 1.1 kloc
  * <details><summary>contains:</summary> extraction/Extraction.v,
    extraction/ExtrHaskellZInteger.v, extraction/ExtrHaskellZInt.v,
    extraction/ExtrHaskellNatNum.v, extraction/ExtrHaskellString.v,
    extraction/ExtrOcamlZInt.v, extraction/ExtrOcamlNativeString.v,
    extraction/ExtrOCamlFloats.v, extraction/ExtrOcamlNatBigInt.v,
    extraction/ExtrOcamlZBigInt.v, extraction/ExtrOcamlNatInt.v,
    extraction/ExtrHaskellNatInt.v, extraction/ExtrOcamlString.v,
    extraction/ExtrHaskellZNum.v, extraction/ExtrOcamlChar.v,
    extraction/ExtrOCamlPArray.v, extraction/ExtrOcamlBasic.v,
    extraction/ExtrOcamlIntConv.v, extraction/ExtrOCamlInt63.v,
    extraction/ExtrHaskellBasic.v,

>     extraction/ExtrHaskellNatInteger.v</details>

Author: I think we shouldn't make primitive depend on extraction. Most uses of primitive I'm aware of don't depend on extraction. We could probably just split `extraction` into `extraction` and `extraction-primitives` though.

>     extraction/ExtrHaskellNatInteger.v</details>

Reviewer 7: I think the correct granularity here is to have standalone extraction-base (should not depend on much, just the ml code and the extraction commands, should work with noinit), and then have a bunch of extraction-foo packages that depend on foo and extraction-base, and maybe extraction loads all of them?

    coq-stdpp. Namespace `Coq.Sets`.
  * 22 files, 3.6 kloc
  * <details><summary>contains:</summary> Sets/Partial_Order.v,
    Sets/Multiset.v, Sets/Uniset.v, Sets/Relations_2.v, Sets/Image.v,
    Sets/Relations_3.v, Sets/Relations_1.v, Sets/Classical_sets.v,
    Sets/Finite_sets.v, Sets/Permut.v, Sets/Integers.v,
    Sets/Finite_sets_facts.v, Sets/Powerset_facts.v, Sets/Powerset.v,
    Sets/Powerset_Classical_facts.v, Sets/Ensembles.v,
    Sets/Relations_1_facts.v, Sets/Constructive_sets.v,
    Sets/Infinite_sets.v, Sets/Relations_2_facts.v,
    Sets/Relations_3_facts.v, Sets/Cpo.v</details>
* coq-roots-sorting
  * The Coq standard library, sorting algorithms
  * Axiomatizations of sorts. Namespace `Coq.Sorting`.
  * 10 files, 4.1 kloc
  * <details><summary>contains:</summary> Sorting/CPermutation.v,
    Sorting/Permutation.v, Lists/SetoidList.v, Sorting/Sorting.v,
    Sorting/Heap.v, Sorting/PermutEq.v, Lists/SetoidPermutation.v,
    Sorting/Mergesort.v, Sorting/PermutSetoid.v,
    Sorting/Sorted.v</details>
* coq-roots-orders-ex
  * The Coq standard library, sorting algorithms
  * Instances of order structures from
    coq-roots-structures. Namespace `Coq.Structures`.

>   * 10 files, 2.5 kloc

Author: My understanding is that orders-ex are instances of order. It then make sense to have them depend on both the order structure and all things being instanciated. But I may be wrong.

    FSets/FSetInterface.v, FSets/FSetWeakList.v, FSets/FSetAVL.v,
    FSets/FSetCompat.v, FSets/FSetDecide.v, FSets/FSetList.v,
    FSets/FSetProperties.v, FSets/FMapList.v, FSets/FMapFacts.v,
    FSets/FSetPositive.v, FSets/FMapFullAVL.v,
    FSets/FSetToFiniteSet.v, FSets/FMapAVL.v, FSets/FMapPositive.v,
    FSets/FSets.v, FSets/FMaps.v, FSets/FSetFacts.v,
    FSets/FSetEqProperties.v, FSets/FSetBridge.v,
    FSets/FMapWeakList.v, MSets/MSetList.v, MSets/MSetDecide.v,
    MSets/MSets.v, MSets/MSetWeakList.v, MSets/MSetGenTree.v,
    MSets/MSetAVL.v, MSets/MSetInterface.v, MSets/MSetFacts.v,
    MSets/MSetProperties.v, MSets/MSetRBT.v, MSets/MSetPositive.v,
    MSets/MSetToFiniteSet.v, MSets/MSetEqProperties.v</details>
* coq-roots-compat
  * The Coq standard library, compatibility files
  * Some compatibility files with older versions. Namespace `Coq.Compat`.
  * 5 files, 0.1 kloc
  * <details><summary>contains:</summary> Compat/AdmitAxiom.v,
    Compat/Coq817.v, Compat/Coq819.v, Compat/Coq818.v,
    Compat/Coq820.v</details>
* coq-roots-derive
  * The Coq standard library, derive
  * Namespace `Coq.derive`
  * 1 file, 0.0 kloc
  * <details><summary>contains:</summary>

>     theories/derive/Derive.v</details>

Author: In any case, we should refrain putting ML code into the library. But since it has no dependencies and is a single line file, it could remain in the stdlib.

* coq-roots-rtauto
  * The Coq standard library, rtauto
  * Namespace `Coq.rtauto`
  * 2 files, 0.8 kloc
  * <details><summary>contains:</summary> rtauto/Bintree.v,

>     rtauto/Rtauto.v</details>

Reviewer 7: Are tauto and rtauto going to be merged some day?

    Wellfounded/Wellfounded.v, Wellfounded/Union.v,
    Wellfounded/Transitive_Closure.v, Wellfounded/Well_Ordering.v,
    Wellfounded/Inverse_Image.v, Wellfounded/Disjoint_Union.v,
    Wellfounded/Lexicographic_Product.v,
    Wellfounded/Lexicographic_Exponentiation.v</details>
* coq-roots
  * The Coq standard library
  * A metapackage for all coq-roots-* packages constituting
    the historical Coq standard library, that used to be shipped with Coq.
  * 508 files, 183.4 kloc

theories/ssr and theories/ssrmatching should go back into mathcomp
(either in coq-mathcomp-ssreflect or in a new coq-mathcomp-ssr
containing only that).

### Documentation

Need to document the above split in packages (could help structure the doc).

## Maintainers

We should do an open call for maintainers. Current maintainers (stdlib
team from Coq on github) will become roots maintainer, they can of
course opt out at any time.

# Drawbacks

The stdlib will no longer come bundled with Coq. It will however be
just as easy to install via the usual package managers. More notably
it will remain an integral part of the Coq Platform.

# Alternatives

> The technical changes proposed here can stand on their own merits, and can be adopted without ratifying the perspective presented for motivating them. Further, each of the three splits discussed under [Packages](https://github.com/proux01/ceps/blob/boost-stdlib-dev/text/083-boost-stdlib-dev/text.md#packages) could be performed without the other ones, with or without the broader vision that led to them.

Author: The sentence there reads "each requiring the previous one", which seems contradictory?

Reviewer 3: Yeah, it does. Could you explain are the requirements because of which 2 cannot happen without 1 and 3 cannot happen without 2?

Author: Well, 2. in https://github.com/proux01/ceps/blob/boost-stdlib-dev/text/083-boost-stdlib-dev/text.md#packages is exposing the split so it need to exist in the first place (that is 1.) and 3. is repo split so hard to do without exposing that split (2.)

1. The logical structure of the library could be split into packages, enforced by the build system, without user-visible structural changes or renaming of stdlib.
2. The standandard library could be moved to its own repository without changing its user-facing relationship with Coq and non-stdlib packaes.

> 3. Specific parts of the standard library could be spun out to their own repositories without changing the rest.

Author: As a side note, this of course only holds for leafs (the dualrepo would also make it possible for roots, not that we would specifically want it though).

Reviewer 3: We could do this at any level using a submodule. I'm not suggesting that we want to right now, but it is absolutely an option.

> Out of these, 1. has the advantage that it retains most of the technical refactoring done for this CEP, and achieves some of the technical goals listed here such as allowing users to `Require` only a known subset, while minimizing disruption to users in the face of multiple possible futures for the standard library more generally. In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.

Author: Could you list them [some of the technical goals] precisely?

Reviewer 3: The most important ones to me are

- Making it clear what files are fair game to Require in each part of stdlib
- Making it clear what will be brought in by a Require of some stdlib file
- While this CEP makes no stability promises for these boundaries, at least it gives names for the components so we can talk about how stable or not they are, and specifically whether we may want to deprecate one or another component

Author: Thanks, integrated to the CEP (feel free to further edit if I made something wrong)

> Out of these, 1. has the advantage that it retains most of the technical refactoring done for this CEP, and achieves some of the technical goals listed here such as allowing users to `Require` only a known subset, while minimizing disruption to users in the face of multiple possible futures for the standard library more generally. In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.

Author: Could you elaborate on those "disruption" and how problematic they look to you?

Reviewer 3: I think distribution packaging is an area with substantial risk; the comments of SnarkBoojum above make me more optimistic about Debian specifically and perhaps that I can follow the same example for Arch and Alpine but I am still somewhat dreading it. My concerns here would be substantially alleviated by a working prototype, or even just making one an explicit launch requirement.

Decoupling the releases of stdlib and Coq could lead to dependency or compat issues affecting them; again it's possible to avoid that but keeping everything together seems like the most reliable way.

Simultaneously, having stdlib need to work with multiple versions of Coq seems like it would further increase rigidity in stdlib. Again, this could be mitigated with careful planning to be done, but I do not miss what we used to do in fiat-crypto in a similar situation.

There are probably a bunch of people with scripts that build coq and code that uses it from git, all of these will need to change if we end up with independent repos. I don't think this should drive our decision, but it does increase the cost of the change.

 (When I wrote this comment, I also thought that having each package have its own release schedule was still on the table.)

Reviewer 3: Thank you for integrating these, I totally should have.

> Out of these, 1. has the advantage that it retains most of the technical refactoring done for this CEP, and achieves some of the technical goals listed here such as allowing users to `Require` only a known subset, while minimizing disruption to users in the face of multiple possible futures for the standard library more generally. In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.

Author: [quoting precisely because the line is way too long to attach comments to a specific part]

> In particular, keeping the boundaries between subparts of the standard library internal to Coq as opposed to exposing them to users would simplify later changing them.

That's a good argument against exposing the package split (point 2. in https://github.com/proux01/ceps/blob/boost-stdlib-dev/text/083-boost-stdlib-dev/text.md#packages ).

My feeling is that you are strongly opposed to the (somewhat orthogonal) dualrepo choice. What are the concrete \
arguments behind that position, I can see

* monorepo is a way to strongly guarantee no additional dependency can be added

are there any other?

Reviewer 3: The key advantage of the monorepo is the one that you already listed in the CEP: it drastically simplifies the logistics of making a change that affects multiple parts (compared to these parts being in separate repositories). A special case of this is that a monorepo makes it easy to ensure internal compatibility by testing every PR in a single CI. Making changes that (conceptually simultaneously) touch coqc, stdlib, and CI developments is a key part of my vision for how to evolve this ecosystem, so I am indeed setting a higher bar for splitting the repo.

While tooling for different workflows within a monorepo can go a long way, I am yet to see a concept for smooth and reliable coordination of changes across different repositories. Overlays as in Coq are very useful given the multirepo reality of Coq CI but they have merge races and pull-request toil (and delays), pinned submodules as in fiat-crypto act as a monorepo with an extra step when everything is moving smoothly and cause an effective freeze of the subrepo code otherwise, maintaining a stable interface across versions like Coq currently strives for is already very limiting on what kind of changes Coq developers can make and I worry it would be even more limiting if the stdlib plugins became a part of that standard interface. Maintaining compatibility in a multirepo environment is also a constant chore in BoringSSL which I sometimes contribute to: there is a porting-duty rotation between the maintainers, and big changes are still tested with supported users before merging anywhere. Basically, every arrangement where a library actually supports its users requires adapting these users to changes at some point, and being able to do this atomically in a single commit in the single repository is the least painful mechanism I have seen, by far.

For a monorepo to make sense, the maintainers of that repository have to have working baseline agreement about what to do with it and mutual trust that only changes that are in line with this agreement are merged. IMO the required consistency bar for is rather low, but it is not zero -- this is why I focus on process questions in #86.  Agreeing to disagree can still work well, and it is not hard to imagine e.g. separate agreed-upon and perhaps even coqbot-enforced rules for different CODEOWNERS. On the other hand, treating some piece as an external dependency with an independent release schedule and compatibility story seems to inherently come with human toil for each coupled change, so I would explore this option only when synchronized development is known to be substantially at odds with the goals of one project or another. I get the sense that you probably do have a vision for stdlib that is at odds with it remaining a part of Coq, but I have only a vague sense for what it is, so it cannot convince me to give up the monorepo.

And as a user of coqc+plugins+stdlib, I count them as "one dependency" exactly because I trust the maintainers of these parts to coordinate effectively enough between themselves that I never have to think about compat between them. (Taking this view to a silly extreme, Debian is "one dependency" too, but it is not the same dependency as Coq, and trying to work with both of them is indeed constraining for fiat-crypto).

Author: Thanks for the explanation. So the root of our disagreement seems to be a strong "philosophical" opinion about monorepo vs. distributed developments with version management. Some people tend to think that repository boundaries should reflect human team boundaries, whereas you'd prefer something wider.

[off topic but maybe interesting]
About the CI of Coq: it seems we also have a quite different approach of it. You seem to see it as some approximation of your monorepo dream. My dream is rather an open world where everyone can maintain its own Coq development without any need to be in the CI, only by easy adaptation to each new Coq / other-dependency release, following its changelog and warning / errors.

You seem to expect a lot from the CI, whereas IMHO, the CI is only a kind of expansive test suite and developers of developments in the CI should only expect:
* less Coq bugs on their dev since CI should catch them early;
* getting involved early in discussions for breaking changes when such changes are needed.

in exchange of having to constantly remain compatible with master.
(Plugins are a particular case where we commit to provide overlays, due to the unstability of the API.)
[/off topic but maybe intresting]

More precised answer to detailled points:

> Overlays as in Coq [...] have merge races

This is theoretically true. However, in my experience, the few merge races I did experience in Coq where never related to overlays. So I wonder how practical the issue is.

> and pull-request toil (and delays)

Right, although again the comment mostly make sense when both sides of the overlay are done by the same "team". Those costs are also a good way to get an idea of how, and at which cost, external devs can be adapted.

> pinned submodules as in fiat-crypto act as a monorepo

Ironically enough, as a Coq developper, my feeling is that those are responsible for a large part of the "overlay toil". Again, only my personal experience, not a fact.

> I get the sense that you probably do have a vision for stdlib that is at odds with it remaining a part of Coq

If by "part of Coq" you mean "same repo", then the main technical point of this CEP is the dualrepo choice, so yes of course. If you mean something else, please clarify.

In any case, it seems we mostly share the same vision of an ideal stdlib. It also seems that the current stdlib should remain usable by all its current users (whether in CI or not) without crazy amounts of effort. Whether the two goals are compatible remains unclear to me.

Reviewer 3: > However, in my experience, the few merge races I did experience in Coq where never related to overlays.

Okay, maybe it's not specifically overlays at fault here. But the loose coupliing of CI developments is definitely a major cause behind CI failures: 4 out of 5 last non-cancellation failures look like an atomically-merge-when-tests-pass policy would have prevented them: https://github.com/coq/coq/runs/23898721987, https://github.com/coq/coq/runs/23874454446, https://github.com/coq/coq/runs/23874454446, https://github.com/coq/coq/runs/23874782784,https://github.com/coq/coq/runs/23874833864

> Those costs are also a good way to get an idea of how, and at which cost, external devs can be adapted.

For stdlib changes, IME the costs of actually adapting the developments are a tiny fraction of the coordination overhead of getting the right person to look at the change and click merge because it is too boring to look at. I think it has happened to me a grand total of once that feedback from a adaptation PR led to a change in the stdlib work.

> Ironically enough, as a Coq developper, my feeling is that those are responsible for a large part of the "overlay toil". Again, only my personal experience, not a fact.

I mean, I'm not saying the submodules are great, just the alternative would be fiat-crypto being red as often as other large multirepo developments. And the particularly slow and painful cases I've witnessed were exactly because some non-coqdev-related codebase-synchronization issues needed to be resolved before the coqdev patch could be accepted.

> Whether the two goals are compatible remains unclear to me.

Which goals are you referring to?

Author: > 4 out of 5 last non-cancellation failures [...]

Those are not "real failures" but the consequence of synchronous overlay merge being indeed non atomic. Synchronous overlays are only a module thing. Since modules would remain in Coq repo, the issue wouldn't affect the library repo. So we	can either consider the	issue orthogonal to the current CEP or even the CEP being an improvement here, since it releases the library from all issues linked to synchronous overlays.

> I'm not saying the submodules are great, just the alternative would be fiat-crypto being red

[off topic]
I really have the feeling that coq-util / bedrock2 / rupicola / fiat-crypto / ... should live in a single repo.
From my outsider point of view, their team seem to have a pretty strong intersection.
[/off topic]

> > Whether the two goals are compatible remains unclear to me.

> Which goals are you referring to?

The ones from the previous sentence:
* evolve the current stdlib to make it truly "standard" / state of the art
* not dramatically breaking every uses

Anyway, that's definitely beyond the scope of the current CEP.

Reviewer 3: I personally agree about fiat-crypto and its dependencies.

Reviewer 3: > Synchronous overlays are only a module thing

I agree currently the two are associated, but are you sure this is fundamental? The sense I get is that the issues currently affecting ML plugins are fundamentally about interfaces that have not been designed for general consumption ending up at a repository boundary, regardless of implementation language. And the interfaces between ML and Gallina/Ltac code for stdlib plugins seem very much like that for me. Another experience confirming this has been maintaining a [copy of non-ML parts of nsatz](https://github.com/mit-plv/fiat-crypto/commits/5563f0925ec0187b9927dbc21b1eab76f15ac7f3/src/Algebra/Nsatz.v) in fiat-crypto.

Reviewer 8: I'm not seeing anything interesting in that commit history, what am I missing?

Reviewer 3: Yeah, the history doesn't really tell a story on its own, and I don't remember well enough to recount it play-by-play. But my takeaway was that a change (making nsatz good to use for custom fields) that would have one commit (or perhaps two for external compat) in a monorepo dragged out over years. This is not purely a technical point on the monorepo/dualrepo choice, as another important factor was slowness and uncertainty of contributing to Coq regardless of synchronization, but nevertheless, it is a negative experience with maintaining Ltac code that wraps an ML plugin in Coq.

Author: I'm not sure what you mean by fundamental. Synchronous overlays are needed for ML code because the ML API is unstable.

In the context of this PR, maybe we should check (both in Coq and the library) that it is properly documented that synchronous overlays are only acceptable for ML code, not for anything else (meaning synchronous overlays would be unacceptable in the library repo).

[off topic]
I agree that due to this, maintaining a ML plugin is bound to be, to some extent, an uncomfortable experience, forcing you to either be compatible with a reduced set of Coq versions or have some mechanism for conditional compilation of multiple plugin versions.
[/off topic]

Reviewer 3: > Synchronous overlays are needed for ML code because the ML API is unstable

> meaning synchronous overlays would be unacceptable in the library repo

I mean that the the Ltac/Gallina APIs for stdlib plugins are no better than the ML ones (it's not about the language!), so freezing them by leaving them at the repository boundary seems unwise.

For spinning out parts of the standard library, one possible heuristic to use is that components with dedicated maintainers could become independent at the initiative of these maintainers. On the other hand, if a part of the standard library is to be separated out without a new maintainer stepping up, this could be handled as a deprecation and removal, not a refactoring.

Other split in multiple repositories would be possible, could still
happen in the future, maybe not needed now. The extreme opposite would
be to lean toward a huge monorepo with everything from the platform,
probably not very practical, and wouldn't bring much since the
platform already ensures uniform releases bundling everything.
To our knowledge the monorepo approach has exactly one advantage:
it is easier to backport changes from one package to an upstream package.
It also has one major constraint (considered an advantage by some and a drawback by others): it makes it harder to have the following dependencies:
`coq-roots-A <- coq-extralib-B <- coq-roots-C`. 

> The most concrete technical advantage of moving stdlib to its own repo, the ability to subscribe to stdlib changes, could also be achieved using a coqbot feature that @-s a specific team of stdlib subscribers.

Author: Could you elaborate on which objectives this does or does not fullfil? I can see as fullfiled:

* enabling stdlib contributors to get only notifications about the library (and not the other Coq PRs)

and as not fullfilled:

* not having lot of OCaml code in the library repo
* not practically needing to compile Coq master to contribute to the library
* avoiding the heavy reviewing process of Coq
*  keeping all doors open with respect to "Enabling the development of an evolved or different official "stdlib""
* "Being sure that users are properly discouraged to rely on parts of the stdlib that are considered subpar, especially when recommended alternatives are available through the Coq Platform."
* enabling the use of Coq without the stdlib (for instance to run the test-suite)
* enabling more freedom on the release process of the library

Reviewer 3: Thank you for this list of goals, somehow it seems much clearer to me than the version in the CEP.

> - not having lot of OCaml code in the library repo

I do not understand why you find this desirable, and I am worried about it. The stdlib plugins are one of the strongest advantages of Coq IMO, and I want to make sure we don't make improving them harder.

> - not practically needing to compile Coq master to contribute to the library

I had not realized this was a goal, I agree it would be nice to have. Yet compiling Coq is such a tiny fraction of the pain of compiling the CI and its dependencies that I am not sure that we should make any decisions based on this without a vision for CI.

> - avoiding the heavy reviewing process of Coq

I agree that the unavailability of reviewers is a problem, but I also find code review valuable. I there anything else you definitely want to see changed?

> - keeping all doors open with respect to "Enabling the development of an evolved or different official "stdlib""

I don't see how this needs any action, the doors are already open.

> - "Being sure that users are properly discouraged to rely on parts of the stdlib that are considered subpar, especially when recommended alternatives are available through the Coq Platform."

We have warnings for that, and we can and have removed content that we don't recommend. I don't think we need another solution.

> - enabling the use of Coq without the stdlib (for instance to run the test-suite)

I agree that this is desirable, but the test suite is such a small fraction of the CI that I am not sure whether this matters (and the dune-based plan seems more appealing to me as long as coq itself is built with dune).

> - enabling more freedom on the release process of the library

Again, I am worried that exercising this freedom would put stress on the plugins that I use Coq for.

Reviewer 3: > We don't want to make it look like mastering OCaml should be anyhow needed to contribute to the library.

Do you think there is a reason why this couldn't be straightforwardly clarified through documentation?

> The future is probably more in the use of dedicated metalanguages like metacoq, Ltac2 or elpi.

Maybe, but I don't see how these new languages would help specifically the stdlib plugins. E.g. elpi has "Lambda" in the name, and lia/field/nsatz do not manipulate any lambdas. Sure, there is some maintenance for with the ML api getting new fancier universes and so on, but constructing application nodes surely isn't changing so unstable that it needs to be avoided. I also expect multiplicative costs with both number of repositories and number of languages, so I would expect we'd do a CEP and consider alternatives for adding any of these to the standard library.

> What makes plugin development hard is the unstable, more or less documented, API, much more than their interaction with the library. So keeping them in Coq repo shouldn't be much trouble.

I agree this is a problem at least for some plugins, but it is not the problem that I was trying to point out. Each plugin exists as a combination of a certificate generator written in ML (perhaps later in another language) and a certificate checker written in Gallina. The interface between them isn't designed for wider consumption, and I have found the documentation to be wrong when I have tried relying on it. Having that interface straddle repositories with separate release schedules would require essentially any improvement to it to have a deprecation delay. This seems quite important to avoid.

> Imagine you're a newcomer and want to add some list lemma, you certainly don't want to compile Coq master and you'd very probably don't have to run any of the CI targets locally since it would just be green online.

Sadly, no. It is especially the simple stuff for which a large part of the effort is multirepo coordination overhead. Often it's just because morally equivalent lemmas with the same name already exist downstream, differing perhaps in names of binders only. [nth_error](https://github.com/coq/coq/pull/18563) was basically as easy as it gets, and I don't want to imagine doing it with web-interface-driven CI.

> No, imagine we choose to go for the "different official "stdlib"" path (I'm not saying we should, I don't know), then we would still have to move the current library to keep maintaining it, so we would be back to the beginning, with that door remaining closed.

Even a well-resourced effort for a total makeover could happen in the same repo. I think what you are saying is that if we want to deprecate the stdlib without porting users, moving it to a separate repo would be the way to do it. Perhaps that's right, but I don't want us to take that path, and making way for it isn't compelling as motivation for this CEP.

> I tend to think that displaying packages, to a proper library, on an equal footing with the alternatives, would help, compared to one of the alternatives being forced the default.

Yeah, we disagree here. IMO it is on us as stdlib maintainers to pick a default and own it. I am especially wary of throwing out old theory/plugins/languages without a replacement that is proven on the same tasks.

> At the very worst, we could imagine schemes were the plugin would have to run alternative code for two different versions of the library, but that need will likely arise very rarely.

How do you imagine this would work?

Reviewer 3: > plugins like paramcoq tend to get quite a lot of overlays

Paramcoq seems vastly more complex in its use of the Coq API than nsatz and micromega to me.

> making a few changes to the micromega plugin in order to make it easily and effectively reusable in a foreign library proved pretty easy and stable enough to not require any change over many Coq versions. So I'm not really worried here, quite the contrary.

My different experience is also with micromega: https://github.com/coq/coq/issues/16862, https://github.com/coq/coq/pull/18730#issuecomment-1978358197.

> [on deprecating stdlib] I understand you are strongly against this possibility, but you cannot deny that not enabling it is keeping doors closed.

But I will deny it: nothing prevents one from making a PR to deprecate any fraction of the stdlib where it sits today. Or is the difference about who would review that PR and how we would expect them to respond?

> "if (new_gallina_constant is present) then new_feature"

Thanks, I agree this could work.

Author: > [on deprecating stdlib]

I fail to understand why you keep bringing "deprecating stdlib" whereas there is no such thing in the CEP.
